python/TestHarness/JobDAG.py

```diff
@@ -128,26 +128,26 @@ def _doSkippedDependencies(self):
 
     def _doRaceConditions(self):
         """ Check for race condition errors within in the DAG"""
-        # Create a clone of the DAG so we can be destructive with it (we need to
-        # simulate running each job to completion)
-        mutable_dag = self.__job_dag.clone()
-        while mutable_dag.size():
-            output_files_in_dir = set()
-            concurrent_jobs = mutable_dag.ind_nodes()
-            for job in concurrent_jobs:
-                if job.isFinished():
-                    continue
-                tester = job.getTester()
-                output_files = tester.getOutputFiles()
-                if len(output_files_in_dir.intersection(set(output_files))):
-                    # Break out of this loop and set every job involved as a failure
-                    # for race conditions
-                    for this_job in concurrent_jobs:
-                        this_job.setStatus(this_job.error, 'OUTFILE RACE CONDITION')
-                    break
-                output_files_in_dir.update(output_files)
-            for job in concurrent_jobs:
-                mutable_dag.delete_node(job)
+        # Build output_file in relation to job dictionary
+        output_to_job = {}
+        for job in self.__job_dag.topological_sort():
+            if job.getRunnable() and not job.isFinished():
+                for output_file in job.getOutputFiles():
+                    output_to_job[output_file] = output_to_job.get(output_file, set([]))
+                    output_to_job[output_file].add(job)
+
+        # Remove jobs which have accurate dependencies
+        for outfile, job_list in output_to_job.iteritems():
+            for job in list(job_list):
+                for match_job in self.__job_dag.all_downstreams(job):
+                    if match_job in job_list:
+                        job_list.remove(match_job)
```

Comment:
Good! This looks like the correct algorithm.

Reply:
But it also is nested deep in for loops :/ I was honestly hoping you had some magic python up your sleeve to beautify this!

Reply:
If you use sets, you can remove the inner loop directly. Look, you are just subtracting matching jobs there, so it's just `job_list = job_list - match_jobs`.

Reply:
@permcody I could do this... but then I would have to store the jobs contained in output_to_job dictionary as a set, instead of a list (or add more lines to convert it later).

What we talked about before, using sets might be slower, and will allow syntax mistakes like this:
https://github.com/idaholab/moose/blob/380145aacd2d1d2f32485bb2cc16b497fc0633ad/test/tests/mesh/adapt/tests#L5

I like the idea of sets. But I thought we agreed upon lists...?

Reply:
Stick with lists, I bet the sets are slower (for small sets)

Reply:
OK to clarify.... Sets are always slower than lists, but your algorithm as it stands is `O(m * n * o)` but by using lists you could make it `O(m * (n + o))`. However, the overhead for these small sets would make the second complexity larger than the first.

Reply:
Thats odd, I went ahead and used sets just for curiosity's sake, removed the inner loop. But then received a race condition with ./run_tests -i duplicate_outputs_ok with tests i and j.

I'll use lists as we discussed.

The error was strange:
```pre
    FAIL tests/test_harness.h FAILED (OUTFILE RACE CONDITION)
    FAIL tests/test_harness.i FAILED (OUTFILE RACE CONDITION) [skipped dependency]
```

Race condition with a skipped dependency caveat... Ummm...

Reply:
Did you use sets for both matched_job and job? If not the subtraction operator might still work but it might fail. You need two sets.